plugins/janus_recordplay.c

```diff
@@ -448,6 +450,80 @@ typedef struct janus_recordplay_session {
 static GHashTable *sessions;
 static janus_mutex sessions_mutex = JANUS_MUTEX_INITIALIZER;
 
+/*! \brief Helper function to filter and map \ref janus_sdp_attribute list to
+ * (char*)id:(char*)info extmap hash table
+ * \param extmaps Hash table to contain filtered and mapped extmaps
+ * \param attributes List of sdp media attributes
+ */
+static void janus_recordplay_map_extmaps(GHashTable *extmaps, const GList *attributes) {
+	if(extmaps == NULL)
+		return;
+	g_hash_table_remove_all(extmaps);
+	for(; attributes != NULL; attributes = attributes->next) {
+		janus_sdp_attribute *attr = (janus_sdp_attribute *)attributes->data;
+		if(!strcasecmp(attr->name, "extmap")) {
+			size_t id_len = strcspn(attr->value, " /");
+			if(!id_len || id_len == strlen(attr->value)) {
+				JANUS_LOG(LOG_WARN, "Extmap 'extmap:%s' is malformed\n", attr->value);
+				continue;
+			}
+			g_hash_table_insert(extmaps, g_strndup(attr->value, id_len), g_strdup(attr->value+id_len));
+		}
+	}
+}
+
+/*! \brief Helper function to copy extmaps from one table to other
+ * \param dst Destination table
+ * \param src Source table
+ */
+static void janus_recordplay_copy_extmaps(GHashTable *dst, GHashTable *src) {
+	if(dst == NULL)
+		return;
+	g_hash_table_remove_all(dst);
+	if(src == NULL)
+		return;
+	GHashTableIter iter;
+	char *ext_id, *ext_info;
+	g_hash_table_iter_init(&iter, src);
+	while(g_hash_table_iter_next(&iter, (gpointer *)&ext_id, (gpointer *)&ext_info))
+		g_hash_table_insert(dst, g_strdup(ext_id), g_strdup(ext_info));
+}
+
+/*! \brief Helper function to update extmap attributes in media
+ * \param media Parsed SDP mline
+ * \param extmaps Table of extmaps to add to media
+ */
+static void janus_recordplay_regenerate_extmaps(janus_sdp_mline *media, GHashTable *extmaps) {
+	if(media == NULL || extmaps == NULL)
+		return;
+	/* Consolidate extmaps */
+	GHashTable *mapped = g_hash_table_new_full(g_str_hash, g_str_equal, g_free, g_free);
+	janus_recordplay_map_extmaps(mapped, media->attributes);
+	GHashTableIter iter;
+	char *ext_id, *ext_info;
+	g_hash_table_iter_init(&iter, extmaps);
+	while(g_hash_table_iter_next(&iter, (gpointer *)&ext_id, (gpointer *)&ext_info))
+		g_hash_table_insert(mapped, g_strdup(ext_id), g_strdup(ext_info));
+	/* Remove existing extmaps */
```

Comment:
I think this is quite wrong: generate offer adds a `mid` extmap, that you'd be removing here. The only thing that you should be doing here is possibly see if an attribute you're adding now already exists (maybe with a different ID), or if there are ID conflicts, so those are the only ones you should remove IMO.

Reply:
Generated extmaps are merged with media attributes into temporary `merged` hashmap, with former having priority, so this would not remove `mid` extmap as it would be present in `merged` hashmap and re-added further down. However, upon inspecting this part of code, it is not very clear how it does what it does, and has some possible edge case errors, so i will rewrite it.